src/services/exportInfoMap.ts

```diff
@@ -445,15 +447,30 @@ export function forEachExternalModuleToImportFrom(
     }
 }
 
-function forEachExternalModule(checker: TypeChecker, allSourceFiles: readonly SourceFile[], excludePatterns: readonly RegExp[] | undefined, cb: (module: Symbol, sourceFile: SourceFile | undefined) => void) {
-    const isExcluded = excludePatterns && ((fileName: string) => excludePatterns.some(p => p.test(fileName)));
+function forEachExternalModule(checker: TypeChecker, allSourceFiles: readonly SourceFile[], excludePatterns: readonly RegExp[] | undefined, host: LanguageServiceHost, cb: (module: Symbol, sourceFile: SourceFile | undefined) => void) {
+    const realpathsWithSymlinks = host.getSymlinkCache?.().getSymlinkedDirectoriesByRealpath();
+    const isExcluded = excludePatterns && (({ fileName, path }: SourceFile) => {
+        if (excludePatterns.some(p => p.test(fileName))) return true;
+        if (realpathsWithSymlinks?.size && pathContainsNodeModules(fileName)) {
+            let dir = getDirectoryPath(fileName);
+            return forEachAncestorDirectory(getDirectoryPath(path), dirPath => {
+                const symlinks = realpathsWithSymlinks.get(ensureTrailingDirectorySeparator(dirPath));
+                if (symlinks) {
+                    return symlinks.some(s => excludePatterns.some(p => p.test(fileName.replace(dir, s))));

```

Comment:
Is `replace` enough here? If we're on a system with `/`, then `dir` can start with `/` which could match basically anywhere in `fileName`. Maybe this is rare but does it need a `prefix + '/'` check?

Reply:
`dir` is necessarily a prefix of `fileName` because it’s just `getDirectoryPath(fileName)` (repeated some number of times)